cdap-common/src/main/java/co/cask/cdap/common/service/Retries.java

```diff
@@ -0,0 +1,157 @@
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.common.service;
+
+import co.cask.cdap.api.Predicate;
+import co.cask.cdap.api.retry.RetriesExhaustedException;
+import co.cask.cdap.api.retry.RetryableException;
+import com.google.common.base.Supplier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Utilities to perform logic with retries.
+ */
+public class Retries {
+  private static final Logger LOG = LoggerFactory.getLogger(Retries.class);
+  private static final Predicate<Throwable> DEFAULT_PREDICATE = new Predicate<Throwable>() {
+    @Override
+    public boolean apply(Throwable throwable) {
+      return throwable instanceof RetryableException;
+    }
+  };
+
+  private Retries() {
+
+  }
+
+  /**
+   * The same as calling {@link #supplyWithRetries(Supplier, RetryStrategy, Predicate)} where a retryable failure
+   * is defined as a {@link RetryableException}.
+   *
+   * @param supplier the callable to run
+   * @param retryStrategy the retry strategy to use if the supplier throws a {@link RetryableException}
+   * @param <T> the type of object returned by the supplier
+   * @return the return value of the supplier
+   * @throws InterruptedException if interrupted while sleeping between retries
+   * @throws RetriesExhaustedException if the supplier failed in a way that is retryable, but the retry strategy
+   *   aborted the operation. The exception cause will be the original throwable thrown by the callable.
+   */
+  public static <T> T supplyWithRetries(Supplier<T> supplier, RetryStrategy retryStrategy) throws InterruptedException {
+    return supplyWithRetries(supplier, retryStrategy, DEFAULT_PREDICATE);
+  }
+
+  /**
+   * Executes {@link Supplier#get()}, retrying the call if it throws something retryable. This is similar to
+   * {@link #callWithRetries(Callable, RetryStrategy, Predicate)}, except it will simply re-throw any non-retryable
+   * exception instead of wrapping it in an ExecutionException. If you need to run logic that throws a
+   * checked exception, use {@link #callWithRetries(Callable, RetryStrategy, Predicate)} instead.
+   *
+   * @param supplier the callable to run
+   * @param retryStrategy the retry strategy to use if the supplier fails in a retryable way
+   * @param isRetryable predicate to determine whether the supplier failure is retryable or not
+   * @param <T> the type of object returned by the supplier
+   * @return the return value of the supplier
+   * @throws InterruptedException if interrupted while sleeping between retries
+   * @throws RetriesExhaustedException if the supplier failed in a way that is retryable, but the retry strategy
+   *   aborted the operation. The exception cause will be the original throwable thrown by the callable.
```

Comment:
An alternative would be to throw the original exception, with addSuppressed(RetriesExhaustedException("no success after %d retries")). That would ensure that this throws the same exception as the method that is being retried, and therefore backward compatible.